### ISSUE TITLE
Merging Practices

### ISSUE THREAD

**Issue by jrmlhermitte** [scikit-beam/scikit-beam#522]
TLDR; We need developer docs for merging and we also need a developer forum. Everyone okay with google groups?

Should we document some recommended merging practices? Currently, we have no agreed method of merging, and I feel it leads to come commits that could have been perhaps saved in a cleaner way. 

There are two main ways as far as I can see, either:
1. Merge everything as is. This makes code forensics potentially easier and portable.
2. Squash everything and write a better more succinct summary (referencing the PR of the merge) Code forensics in this case could eventually involve external information (i.e. visiting the github PR) but I'd argue that even for the former doing so would also be beneficial

What brought me to write this issue is the following as an example:
```commit 24aad638a426a19d5c9507f59132107636296f2b
Merge: b7e2b94 021eaad
Author: Li Li (omitted email)
Date: Fri Mar 24 15:44:36 2017 -0400

Merge pull request #494 from ordirules/CrossCorrelator
    
New and improved Cross correlator

```
There are a ton of I'd say useless commits that amounted to it (this is mine so I'm allowed to be critical :-D). If you go down the history, there is one commit sandwiched between them that actually contains useful information:
```commit fc23711dd43fa7ed3332ae2c23ff877ffc7c5b7a
Author: Julien Lhermitte (omitted email)
Date: Tue Mar 14 10:20:49 2017 -0400

Incorporated Mark Sutton's changes. Code is about 40% faster.
Changes:
1. Used more efficient indexing tricks to index in images
2. Removed the necessity of expanding the image (fftconvol does this for free)
3. Removed the wrap parameter (since fftconvol expands the image to speed up computations,
wrapping is not possible). This parameter was never used so it is simply ignored.
4. For now, removed the 1D case of CrossCorrelator (to be added in later commit)
5. Modified tests for 2D cross correlator. New code results are identical to old,
with the exception that new code more efficiently wraps the image.
(Basically, previous result was zero padded at edges, this version is not)
6. Removed the memory intensive allocation of arrays for the computations during
the object initialization.
Instead, the arrays are allocated and de-allocated per computation (per bin id).
This is likely close in expense as zero-ing the array, which happened regardless.
```
I would argue that things like this could be condensed into one summary commit with a succinct summary. Worst comes to worst use the reference PR from github for further analysis.

Personally, I like how [dask](http://www.github.com/dask/dask) handles their commits. For the most part, commits from PR's are squashed into a single commit with a nice summary.

We could start by writing out explicit documentation for merging practices as a PR (and perhaps argue in the PR), maybe [here](http://scikit-beam.github.io/scikit-beam/resource/dev_guide/index.html). Also, we probably might need a more free form forum for general discussion of ideas (that don't necessarily need to be tracked as an issue).

I suggest opening a google group for `scikit-beam-dev`. What does everyone think?
@danielballan @tacaswell @licode @CJ-Wright  @stuartcampbell @arkilic @ericdill  (and anyone else)

**Comment by CJ-Wright** [comment]
There is another option, we could make a `news` folder and demand that all PRs include a new file in that folder. When it comes time to release then we can collate that into a `CHANGELOG`. This way we get the bet of both worlds, our git history is the history of the sofware as written but the news/changelog has a nice description of what the PR meant. This is how [xonsh](https://github.com/xonsh/xonsh) does things and it works very well with [rever](https://github.com/ergs/rever)

**Comment by ericdill** [comment]
@jrmlhermitte -- thank you for the careful consideration of this issue. I think that the idea of squashing PRs into a single commit has merit and am generally 👍 on the idea. I know that @tacaswell and I had chatted about the pros/cons of this at some point in the past and I am curious what his thoughts on the matter are now.

To @CJ-Wright 's point of "the history of the software as written", I think that there would be a great deal of value in preserving the git history as developed ***if*** every single commit from every single person contributing to this codebase had a coherent and descriptive commit message. Please note that this is not me throwing stones here as I am as guilty as anyone else for not writing verbose and descriptive commit messages.

Overall I'm more 👍 than 👎. I think this is a healthy thing to discuss and am curious what the others think

**Comment by jrmlhermitte** [comment]
randomly came across this. Honestly, I prefer the 'squash and merge' option. I often find in our repos that many micro commits are for the most part useless and add more noise than benefit.
A PR should not require more granularity, and we well described.

Those are my two cents. Maybe decide on yes or no and just close this issue either way?

**Comment by jrmlhermitte** [comment]
Closing (but please open if you think this is important)
